ios/FluentUI/Bottom Sheet/BottomSheetController.swift

```diff
@@ -0,0 +1,376 @@
+        case .ended, .cancelled, .failed:
+            completePan(with: sender.velocity(in: view).y)
+        default:
+            break
+        }
+    }
+
+    private func translateSheet(by translationDelta: CGPoint) {
+        let maxOffset = expandedOffsetFromBottom + Constants.maxRubberBandOffset
+        let minOffset = collapsedOffsetFromBottom - Constants.maxRubberBandOffset
+
+        var offsetDelta = translationDelta.y
+        if currentOffsetFromBottom <= collapsedOffsetFromBottom || currentOffsetFromBottom >= expandedOffsetFromBottom {
+            offsetDelta *= translationRubberBandFactor(for: currentOffsetFromBottom)
+        }
+        bottomSheetOffsetConstraint.constant = -min(max(currentOffsetFromBottom - offsetDelta, minOffset), maxOffset)
+    }
+
+    private func translationRubberBandFactor(for currentOffset: CGFloat) -> CGFloat {
+        var offLimitsOffset: CGFloat = 0.0
+        if currentOffset > expandedOffsetFromBottom {
+            offLimitsOffset = min(currentOffset - expandedOffsetFromBottom, Constants.maxRubberBandOffset)
+        } else if currentOffset < collapsedOffsetFromBottom {
+            offLimitsOffset = min(collapsedOffsetFromBottom - currentOffset, Constants.maxRubberBandOffset)
+        }
+
+        return max(1.0 - offLimitsOffset / Constants.maxRubberBandOffset, Constants.minRubberBandScaleFactor)
+    }
+
+    // MARK: - Animations
+
+    private func completePan(with velocity: CGFloat) {
+        if abs(velocity) < Constants.directionOverrideVelocityThreshold {
+            // Velocity too low, snap to the closest offset
+            if abs(collapsedOffsetFromBottom - currentOffsetFromBottom) < abs(expandedOffsetFromBottom - currentOffsetFromBottom) {
+                move(to: .collapsed, velocity: velocity)
+            } else {
+                move(to: .expanded, velocity: velocity)
+            }
+        } else {
+            // Velocity high enough, animate to the offset we're swiping towards
+            if velocity > 0 {
+                move(to: .collapsed, velocity: velocity)
+            } else {
+                move(to: .expanded, velocity: velocity)
+            }
+        }
+    }
+
+    private func move(to targetExpansionState: BottomSheetExpansionState, animated: Bool = true, velocity: CGFloat = 0.0) {
+        if animated {
+            animate(to: targetExpansionState, velocity: velocity)
+        } else {
+            bottomSheetOffsetConstraint.constant = -(targetExpansionState == .expanded ? expandedOffsetFromBottom : collapsedOffsetFromBottom)
+            view.setNeedsLayout()
+
+            switch targetExpansionState {
+            case .expanded:
+                delegate?.bottomSheetControllerDidExpand?(self)
+            case .collapsed:
+                delegate?.bottomSheetControllerDidCollapse?(self)
+            }
+        }
+    }
+
+    private func animate(to targetExpansionState: BottomSheetExpansionState, velocity: CGFloat = 0.0) {
```

Comment:
This is really clear to me and kinda awesome. Looking forward to trying it out elsewhere